archive/imo/imo1998_q2.lean

```diff
@@ -0,0 +1,169 @@
+  (A r).card ≤ k * ((fintype.card J) * (fintype.card J) - (fintype.card J)) :=
+begin
+  change _ ≤ k * ((finset.card _ ) * (finset.card _ ) - (finset.card _ )),
+  rw ← finset.off_diag_card,
+  apply finset.card_le_mul_card_image_of_maps_to (A_snd_mem r),
+  rintros ⟨j₁, j₂⟩ hj,
+  have hj' : j₁ ≠ j₂, { simp [finset.mem_off_diag] at hj, exact hj, },
+  rw ← A_snd_fibre_card r hj', apply hk, exact hj',
+end
+
+end
+
+lemma add_sq_add_sq_sub {α : Type*} [ring α] (x y : α) :
+  (x + y) * (x + y) + (x - y) * (x - y) = 2*x*x + 2*y*y :=
+by noncomm_ring
+
+lemma norm_bound_of_odd_sum (x y z : ℤ) (h : x + y = 2*z + 1) :
+  2*z*z + 2*z + 1 ≤ x*x + y*y :=
+begin
+  suffices : 4*z*z + 4*z + 1 + 1 ≤ 2*x*x + 2*y*y,
+  { rw ← mul_le_mul_left zero_lt_two, convert this; ring, },
+  have h' : (x + y) * (x + y) = 4*z*z + 4*z + 1, { rw h, ring, },
+  rw [← add_sq_add_sq_sub, h', add_le_add_iff_left],
+  suffices : 0 < (x - y) * (x - y), { apply int.add_one_le_of_lt this, },
+  apply mul_self_pos, rw sub_ne_zero, apply int.ne_of_odd_sum ⟨z, h⟩,
+end
+
+section
+
+variables [fintype J]
+
+lemma agreement_lower_bound (z : ℕ) (hJ : fintype.card J = 2*z + 1) (c : C) :
+  2*z*z + 2*z + 1 ≤ (finset.univ.filter (λ (x : J × J), r c x.1 = r c x.2)).card :=
+begin
+  let x := (finset.univ.filter (λ j, r c j)).card,
+  let y := (finset.univ.filter (λ j, ¬ r c j)).card,
+  have h : (finset.univ.filter (λ (x : J × J), r c x.1 = r c x.2)).card = x*x + y*y,
+  { simp [← finset.filter_product_card], },
+  rw h, apply int.le_of_coe_nat_le_coe_nat, simp only [int.coe_nat_add, int.coe_nat_mul],
+  apply norm_bound_of_odd_sum ↑x ↑y ↑z,
+  suffices : x + y = 2*z + 1, { simp [← int.coe_nat_add, this], },
+  rw [finset.filter_card_add_filter_neg_card_eq_card, ← hJ], refl,
+end
+
+lemma agreement_lower_bound' (z : ℕ) (hJ : fintype.card J = 2*z + 1) (c : C) :
+  2*z*z ≤ (finset.univ.filter (λ (x : J × J), r c x.1 = r c x.2 ∧ x.1 ≠ x.2)).card :=
+begin
+  let s := finset.univ.filter (λ (x : J × J), r c x.1 = r c x.2),
+  let t := finset.univ.filter (λ (x : J × J), x.1 ≠ x.2),
+  have hs : 2*z*z + 2*z + 1 ≤ s.card, { exact agreement_lower_bound r z hJ c, },
+  have hst : s \ t = finset.univ.diag, { ext, split; intros; finish, },
+  have hst' : (s \ t).card = 2*z + 1, { rw [hst, finset.diag_card, ← hJ], refl, },
+  rw [finset.filter_and, finset.inter_eq_sdiff_sdiff s t, finset.card_sdiff],
+  { rw hst', rw add_assoc at hs, apply nat.le_sub_right_of_add_le hs, },
+  { apply finset.sdiff_subset_self, },
+end
+
+lemma agreement_lower_bound'' [fintype C] (z : ℕ) (hJ : fintype.card J = 2*z + 1) :
+  2*z*z * (fintype.card C) ≤ (A r).card :=
+begin
+  have A_fst_mem : ∀ a, a ∈ A r → prod.fst a ∈ @finset.univ C _, { intros, apply finset.mem_univ, },
+  apply finset.mul_card_image_le_card_of_maps_to A_fst_mem,
+  intros c hc,
+  rw ← A_fst_fibre_card,
+  apply agreement_lower_bound' r z hJ,
+end
+
+end
+
+lemma annoying_lemma_1
```

Comment:
Perhaps there is some more useful name for these than calling them `annoying_lemma_1` and `annoying_lemma_2`? I would find it useful to have a short comment explaining why these lemmas are annoying, why you need them anyway, that sort of thing.

Reply:
You're absolutely right. I have now refactored things slightly and completely removed those names.